### ISSUE TITLE
Add option to specify extra flags for cluster up

### ISSUE THREAD

**Issue by gbraad** [minishift/minishift#1170]
Allow to specify extra options/flags for the startup with `oc cluster up` by adding eg. `--oc-cluster-extra-flags="--service-catalog"`

See also discussion on issue #1137

**Comment by LalatenduMohanty** [comment]
As @jorgemoralespou mentioned in https://github.com/minishift/minishift/issues/1137#issuecomment-315601681
```
For the future, maybe it would be useful to be able to provide additional flags that will get passed "as is" to the oc cluster instantiation in minishift, to be able to early test these functionalities without needing to provide additional capabilities in minishift.

e.g.:

minishift start --openshift-version=v3.6.0-rc.0 --openshift-extra-flags='--service-catalog'
```

**Comment by gbraad** [comment]
suggested by @praveenkumar and me it is better to call this option `--oc-cluster-extra-flags=""`

**Comment by LalatenduMohanty** [comment]
@gbraad IMO it should be `--clusterup-extra-flags` as the command name is `cluster up` and it communicates the purpose well to me.

**Comment by hferentschik** [comment]
>  it should be --clusterup-extra-flags as the command name is cluster up and it communicates the purpose well to me

I agree that it should contain 'clusterup' and I might also change the order a bit _--extra-clusterup-flags_.

**Comment by hferentschik** [comment]
I am not sure whether this is such a good idea though. Would this flag try to filter out flags which can also be specified via the CLI? Or would the idea be to pass on flags as is. In the latter case, what applies the command explicitly specified or the one via --extra-clusterup-flags? The behaviour would need to be deterministic at least. 

Also on the community side, we are able to add flags quite quickly, so I wonder whether it makes sense to  have this additional flag. The situation might be different if we were to say that moving forward we would only use this flag to enable features like service catalog (or metrics, even though too late for it now). In this case, however, we would most likely not add the explicit --service-catalog flag. 

Awe should also think about whether this kind of features are not best supported via the add-on mechanism.

**Comment by hferentschik** [comment]
Another approach is to add extra-clusterup-flags as environment variable (MINISHIFT_EXTRA_CLUSTERUP_FLAGS) and/or in combination with a hidden Viper flag. In this case the features becomes more of a power user feature and gets out of harms way for normal users. It also does not continue to clutter the standard minishift start flags.

I think of all the options mentioned so far, I prefer this option best atm. It is somewhere a middle ground.

**Comment by jorgemoralespou** [comment]
Sounds good to me.

**Comment by gbraad** [comment]
what if we introduced a `--future` flag to enable options that we do not support yet, but want to expose in code we can use a `if future { viper.AddFlag() }` so they do not show up unless `--future` is used pretty much a feature toggle, like python did `import future`

**Comment by gbraad** [comment]
I created a test for this at a personal branch: https://github.com/gbraad/minishift/commit/6d2121b1475e534c4c057bd63652c2a4633d4486

It is not possible to add a top-level flag for `RootCmd`, as they get parsed on the level of `startCmd`. And then also, you can't use a flag to toggle the flagset... So I was only able to do the following:

```
$ MINISHIFT_FUTURE=true minishift start --service-catalog
$ MINISHIFT_FUTURE=true minishift start --extra-clusterup-flags="--service-catalog"
```

This also prevents us from having to find a way to hide the Viper flag, as the future enevironment variables toggles the `future` boolean. And this can actually live on the `RootCmd` if needed, so `if future` can be used broadly... and not just for the `start` subcommand.

**Comment by LalatenduMohanty** [comment]
@gbraad We should try https://godoc.org/github.com/spf13/pflag#FlagSet.MarkHidden 

I am ok with `minishift start --future --clusterup-flags=service-catalog` . It is fine to make both --future and --clusterup-flags hidden.

**Comment by gbraad** [comment]
The problem is that I am not able to use a flag to toggle other flags... The value remained `false`, even when requested earlier.

When used in this way, you can at least show the actual flags when using `MINISHIFT_FUTURE=y minishift -h`. Haven't checked if `MarkHidden` allows you to override the hidden state. But anyways... I would like if `--hidden` would be possible.

**Comment by hferentschik** [comment]
To me the '--future' flags sounds to add even more complexity. Can we not go with the environment variable approach first? This seems to be less invasive and gives us more flexibility to change.

**Comment by LalatenduMohanty** [comment]
@hferentschik as per @gbraad 's current PR `MINISHIFT_FUTURE=y ` is an env variable and then if ot is set the --extra-clusterup-flags will show up. Does that work for you?

**Comment by gbraad** [comment]
As per discussion on #1185 I will at least make the change to `MINISHIFT_ENABLE_EXPERIMENTAL` or similar and harden the check to only accept a small subset of options to enable it.

**Comment by hferentschik** [comment]
Merged via pull request #1185
